Drivers: DAI: Intel: DMIC: Program start symmetrically for PDMx

singalsu opened the pull request:
This PR contains proposed improvements to ACE and cAVS DMIC start sequence to avoid possible PDMx controllers out-of-sync start.

singalsu commented on drivers/dai/intel/dmic/dmic.c:
```diff
@@ -620,15 +620,11 @@ static void dai_dmic_start(struct dai_intel_dmic *dmic)
 	/* Clear soft reset for all/used PDM controllers. This should
-	 * start capture in sync.
+	 * start capture in sync. Side-by-side launch is time critical.
+	 * Not not produce LOG_INF() traces here.
```
Typo, should be "Do not"

lyakh commented on drivers/dai/intel/dmic/dmic.c:
```diff
@@ -552,16 +552,16 @@ static void dai_dmic_start(struct dai_intel_dmic *dmic)
 	/* Initial gain value, convert Q12.20 to Q2.30 */
 	dmic->gain = Q_SHIFT_LEFT(db2lin_fixed(LOGRAMP_START_DB), 20, 30);
 
+	key = k_spin_lock(&dmic->lock);
```
it depends really on what this lock is protecting. Is it protecting only the `.startcount` field? Or can other `dmic` fields (`.gain` and `.gain_coef`) be modified asynchronously too? Since this is a "start" function, I suspect that at this time no other context can try to access the DMIC DAI, so maybe you don't even need to protect the `.startcount`. Or is this protecting against DMIC interrupts triggering during initialisation? But in fact I agree with the comment in the commit message, that "this isn't fixing any existing issue"

ujfalusi replied:
if you protect the `startcount` then you should protect the gain as well, they both used in dai_dmic_gain_ramp() under lock held and then probably the gain_coef as well.

It might be that protecting these variables are not needed at all...

ujfalusi commented on drivers/dai/intel/dmic/dmic.c:
```diff
@@ -620,15 +620,11 @@ static void dai_dmic_start(struct dai_intel_dmic *dmic)
-			dai_dmic_read(dmic, dmic_base[i] + CIC_CONTROL));
-	}
+	for (i = 0; i < CONFIG_DAI_DMIC_HW_CONTROLLERS; i++)
+		dai_dmic_update_bits(dmic, dmic_base[i] + CIC_CONTROL, CIC_CONTROL_SOFT_RESET, 0);
```
Zephyr likes to have { } even for single functions in loop/if cases

ujfalusi commented on drivers/dai/intel/dmic/dmic.c:
```diff
@@ -566,19 +566,14 @@ static void dai_dmic_start(struct dai_intel_dmic *dmic)
-		LOG_INF("dmic_start(), cic 0x%08x",
-			dai_dmic_read(dmic, dmic_base[i] + CIC_CONTROL));
+	for (i = 0; i < CONFIG_DAI_DMIC_HW_CONTROLLERS; i++)
+		dai_dmic_update_bits(dmic, dmic_base[i] + CIC_CONTROL, CIC_CONTROL_SOFT_RESET, 0);
```
put { } around the loop.

singalsu replied:
Is it coding style or a technical thing?

singalsu committed:
Drivers: DAI: Intel: DMIC: Program start symmetrically for PDMx

This change is assumed to fix the random corruption of 4ch
capture for PDM1 channels 3-4 in PTL platform. There are no
solid facts behind this change but assumption that PDMx
controllers are not in sync if the start sequence for PDM1
is further away from PDM0. The PDM0 internal state may be
different from PDM1

The single for loop to handle the CIC and FIR start sequence
is split into two for loops to handle same registers update
tasks symmetrically for all stereo PDM controllers. E.g. two
PDMs for four microphones.

First loop programs the CIC_CONTROL and MIC_CONTROL registers
of the PDMx controllers. These features belong to the CIC block
in DMIC IP. Second loop programs the FIR_CONTROL registers of
the PDMx controllers.

In a stress test of 100 times repeated commands:

arecord -Dhw:0,6 -fS32_LE -r48000 -c4 -d 10 dmic_test_1.wav; \
sleep 0.5; \
arecord -Dhw:0,6 -fS32_LE -r48000 -c4 -d 10 dmic_test_2.wav; \
sleep 1

The corruption occurrence with xt-clang build was e.g. 87/200
fails in one of wav files giving 43.5% occurrence. The test was
done with Zephyr commit fe29c40a9366b5ffdcdd2eac26023ce4502413b1
("llext: add inspection API test suite").

In a gcc build the occurrence of corruption is lower, around 6%
but it is seen that the channels 3-4 pdm1 are swapping randomly.

With this fix the corruption occurred zero times in xt-clang
and gcc builds with same 100 repeats.

Signed-off-by: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>

singalsu commented:
I rebased this patch and dropped commit for spinlock set move. A SOF PR test for this patch runs now in https://github.com/thesofproject/sof/pull/9859 .

singalsu commented:
With the failing version, with xt-clang, the trimmed output looks like this. The -132.2 dB levels are from disconnected microphones in DUT for channels 2 and 4. In the end there is 43.5% failures.

```
       1,    -49.6,   -132.2,    -49.5,   -132.2,       0,        0,      0.0
       2,    -49.2,   -134.9,     -0.9,   -134.9,       1,        1,     50.0
       3,    -49.4,   -135.1,     -0.9,   -135.1,       1,        2,     66.7
       4,    -49.3,   -135.0,    -49.2,   -135.0,       0,        2,     50.0
...
     197,    -49.3,   -132.2,     -0.9,   -132.2,       1,       85,     43.1
     198,    -49.2,   -132.2,     -0.9,   -132.2,       1,       86,     43.4
     199,    -49.6,   -134.9,     -0.9,   -134.9,       1,       87,     43.7
     200,    -49.2,   -133.2,    -49.2,   -133.2,       0,       87,     43.5

```

With gcc build, there is channels swapping happening. It can be seen with channels 3-4 levels . The disconnected level of -73 dB is due to different version of equalizer for gcc build. In the end 5.5% are failed. The channels swaps are not counted in the script.

```

       1,    -47.5,    -73.3,    -73.0,    -47.0,       0,        0,      0.0
       2,    -48.8,    -73.3,    -48.8,    -73.1,       0,        0,      0.0
       3,    -47.3,    -73.1,    -47.1,    -73.1,       0,        0,      0.0
       4,    -47.8,    -73.3,    -73.0,    -47.8,       0,        0,      0.0
---
     197,    -49.7,    -73.2,    -73.0,    -49.7,       0,       10,      5.1
     198,    -49.9,    -73.1,     -1.3,    -73.1,       1,       11,      5.6
     199,    -49.7,    -73.3,    -49.6,    -73.1,       0,       11,      5.5
     200,    -50.1,    -73.2,    -50.1,   -104.4,       0,       11,      5.5

```

singalsu commented:
I lost PTL device while stress testing this patch with xt-clang build. The gcc build test completed for 100 loops without corruptions and channels swaps.
 
```
...
     197,    -49.9,    -73.1,    -50.0,    -73.1,       0,        0,      0.0
     198,    -50.0,    -73.1,    -50.3,    -73.1,       0,        0,      0.0
     199,    -50.1,    -73.3,    -50.4,    -73.1,       0,        0,      0.0
     200,    -50.0,    -73.1,    -50.4,    -73.1,       0,        0,      0.0
```

singalsu commented:
The PTL xt-clang build of this patch passed 500 loops without corruption or channels swap.

```
---
     997,    -49.8,   -132.2,    -50.5,   -132.2,       0,        0,      0.0
     998,    -49.9,   -132.2,    -50.5,   -132.2,       0,        0,      0.0
     999,    -49.9,   -132.2,    -50.5,   -132.2,       0,        0,      0.0
    1000,    -49.6,   -135.2,    -50.3,   -134.5,       0,        0,      0.0
```

singalsu commented:
I rebased and pushed again to restart CI checks, no changes to patch.
